save EVE fatal, stdout, etc to files on persistent storage

gkodali-zededa opened the pull request:
1) Create a per agent debug directory in /persist/agentdebug/ with agentname as the name of directory.
2) Apart from forwarding to cloud, Redirect stdout, sigusr1, sigusr2, fatal stack, fatal reason into seperate files in the above created directory. Latest content should overwrite the old contents in these files. 

Unexpected service crashes will still go to cloud as logs.

Signed-off-by: Gopi krishna Kodali <gkodali@zededa.com>

gkodali-zededa commented:
> To facilitate local (ssh to dom0) developper debugging this makes sense.
> 
> But I think we should structure and name the directories in /persist slightly differently so we can use that layout for additional information.
> 
> I'd suggest something like /persist/agentdebug/agentname/ and in there we can have the following set of files:
> starttime - an empty file which is touched when the agent is started
> stdout - from your code; separate from stderr
> stderr - from your code; will most likely contain panic info/stack trace
> fatal - we can add code for log.Fatal to print the fatal string here
> fatalstack - the stack trace from the log.Fatal
> sigusr1 - the stack traces from SIGUSR1 are written to this file
> sigusr2 - the memory info from SIGUSR2 are written to this file
> 
> The above files will never be removed, but a subsequent write will overwrite the previous content (this not append to it). That makes it easy for a developer to look at these files post a reboot/crash.
> 
> The "starttime" file is critical to determine whether some file was generated by the agent the last time it was started. Thus if the modification time of a file (like fatal or stderr) is older than the modification time of the starttime file, then those are from a previous/old run of the agent.
> That enabled us to automatically extract information from /persist/agentdebug either from the watchdog-report script, or by running some code in device-steps.sh on the way up.
> 
> So I'd suggest three things for this PR:
> 
> 1. Use above naming scheme
> 2. Split stderr and stdout to point at separate files
> 3. Create/touch the starttime file when the agent starts.
> 
> Then we can add the other info in separate PRs.

Makes sense. Will do 1, 2, 3 and update PR.

rvs commented:
Hey @eriknordmark @gkodali-zededa -- I thought we agreed to handle this differently as part of the overall logging pipeline improvement. Why are we still chasing this instead of actually fixing what's broken (the pipeline) @gkodali-zededa ?

eriknordmark commented:
> Hey @eriknordmark @gkodali-zededa -- I thought we agreed to handle this differently as part of the overall logging pipeline improvement. Why are we still chasing this instead of actually fixing what's broken (the pipeline) @gkodali-zededa ?

Because recent information I received that customer success folks are having a hard to to extract this information from the logs even for systems in the lab where they have ssh access. (And I personally have been running with my personal workaround to get reliable logs on the device for months.)
Wishing that the logging pipeline become usable hasn't worked for me, and when I found it isn't' working for CS either I asked Gopi to do this local saving of state to handle the case when we have ssh access. 
Doesn't mean we shouldn't make the logging pipeline reliable and usable end to end, but that seems to still be months away.

gkodali-zededa commented:
> > Hey @eriknordmark @gkodali-zededa -- I thought we agreed to handle this differently as part of the overall logging pipeline improvement. Why are we still chasing this instead of actually fixing what's broken (the pipeline) @gkodali-zededa ?
> 
> Because recent information I received that customer success folks are having a hard to to extract this information from the logs even for systems in the lab where they have ssh access. (And I personally have been running with my personal workaround to get reliable logs on the device for months.)
> Wishing that the logging pipeline become usable hasn't worked for me, and when I found it isn't' working for CS either I asked Gopi to do this local saving of state to handle the case when we have ssh access.
> Doesn't mean we shouldn't make the logging pipeline reliable and usable end to end, but that seems to still be months away.

1) Removed code that re-directs stderr to persist file. Fixed sending unexpected crash traces to cloud as part of a separate PR.
2) Now service stdout/fatal traces/USR1/USR2 traces will go into files in /persist/agentdebug/<agentname>/ directory. This happens in addition to these logs/traces being exported to cloud.
3) Touch files per service to indicate service start time.

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -79,13 +79,17 @@ func handleSignals(sigs chan os.Signal) {
 				log.Warnf("SIGUSR1 triggered with %d stacks", len(stackArray))
 				for _, stack := range stackArray {
 					log.Warnf("%v", stack)
+					// This goes to /persist/agentdebug/<agentname>/stdout file
+					fmt.Printf("%v", stack)
```
I would actually have a separate agentname/sigusr1 file for this to make it easier to find.

gkodali-zededa replied:
ok

gkodali-zededa replied:
done

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -79,13 +79,17 @@ func handleSignals(sigs chan os.Signal) {
 				RebootStack(stacks)
 			case syscall.SIGUSR2:
 				log.Warnf("SIGUSR2 triggered memory info:\n")
+				// This goes to /persist/agentdebug/<agentname>/stdout file
+				fmt.Printf("SIGUSR2 triggered memory info:\n")
```
Ditto separate sigusr2 file.

gkodali-zededa replied:
ok

gkodali-zededa replied:
done

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -317,10 +330,67 @@ func roundToMb(b uint64) uint64 {
+
+// Send application stdout/stderr to this panic file
+func spoofStdFDs(agentName string) *os.File {
+	//filename := fmt.Sprintf("%s/%s", types.PersistDebugDir, panicFileName)
```
Remove?

gkodali-zededa replied:
done

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -317,10 +330,67 @@ func roundToMb(b uint64) uint64 {
+	}
+}
+
+// Send application stdout/stderr to this panic file
```
Update comment?

gkodali-zededa replied:
done

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -317,10 +330,67 @@ func roundToMb(b uint64) uint64 {
+	//filename := fmt.Sprintf("%s/%s", types.PersistDebugDir, panicFileName)
+	agentDebugDir := fmt.Sprintf("%s/%s/", types.PersistDebugDir, agentName)
+	if _, err := os.Stat(agentDebugDir); os.IsNotExist(err) {
+		// Create the agent specific trace directory
```
s/trace/debug/

gkodali-zededa replied:
done

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -317,10 +330,67 @@ func roundToMb(b uint64) uint64 {
+	startTimeFile := agentDebugDir + "/starttime"
+	touchFile(startTimeFile)
+
+	panicOutFile := agentDebugDir + "/stdout"
```
Since this no longer gets panics, a different name for the local variable would be helpful.

gkodali-zededa replied:
done

gkodali-zededa commented:
> I think the stdout, SIGUSR1, SIGUSR2, fatal string, and fatal stack trace can and show go to different files.
> That way if we need to extract them (from the watchdog-report script or post-reboot) it is a lot easier.

done

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -76,6 +80,18 @@ func handleSignals(sigs chan os.Signal) {
 				stacks := getStacks(true)
 				stackArray := strings.Split(stacks, "\n\n")
+
+				sigUsr1File, err := os.OpenFile(sigUsr1FileName, os.O_WRONLY|os.O_CREATE|os.O_SYNC, 0755)
```
I forget whether this does append by default if case you have to specific O_TRUNC to start empty.

eriknordmark replied:
You do.
I don't think you need O_SYNC. So you can use overwriteFile

eriknordmark replied:
Oh - you have multiple writes here.
So can't use overwriteFile

gkodali-zededa replied:
Took care of it

gkodali-zededa replied:
> > Doesn't mean we shouldn't make the logging pipeline reliable and usable end to end, but that seems to still be months away.
> 
> Fair enough @eriknordmark -- I am still unhappy about `spoofStdFDs` though -- the way I read it seems that we're now circumventing the normal stdout/stderr flow that feeds memlogd? How's that supposed to work @gkodali-zededa ?

@rvs Service starts with it stdout, stderr attached to pipes that terminate at containerd-shim.

stdout ----------------------------> containerd-shim-fdX
stderr ----------------------------> containerd-shim-fdY

We make a clone of stdout. New cloned FD is created such that message sent to this cloned FD also go over the same pipe that stdout is attached to.

stdout ----------------------------> containerd-shim-fdX
cloned-stdout --------------------> containerd-shim-fdX
stderr ----------------------------> containerd-shim-fdY

Now we give this cloned-stdout to logrus. Logs go to containerd-shim using cloned-stdout socket.

We then create a file FD and then copy this new file FD (properties) into stdout FD.

stdout ----------------------------> containerd-shim-fdX
cloned-stdout --------------------> containerd-shim-fdX (via logrus)
stderr ----------------------------> containerd-shim-fdY
file-fd ----------------------------> file-on-persist

becomes

stdout ----------------------------> file-on-persist
cloned-stdout --------------------> containerd-shim-fdX (via logrus)
stderr ----------------------------> containerd-shim-fdY

There are other files that we create in persist like fatal-stack, fatal-reason, sigusr1, sigusr2 to which we explicitly write from respective handlers the appropriate messages/traces when relevant events occur.

rvs replied:
Hey @gkodali-zededa -- did you mean:
```
stderr ----------------------------> file-on-persist
cloned-stdout --------------------> containerd-shim-fdX (via logrus)
stderr ----------------------------> containerd-shim-fdY
```

In your last diagram? I don't see where stdout gets redirected to the file in persist in the code (and @eriknordmark seems to agree that it is only stderr that we pipe to files)

gkodali-zededa replied:
My diagram is correct. It is only stdout that we re-direct to file. Look for Dup2 system call, where we copy persist file FD props to stdout FD.
https://github.com/lf-edge/eve/blob/50c4122829be096a762cb7989b5f41dc36052f6f/pkg/pillar/agentlog/agentlog.go#L434

stderr (unexpected crash/panic traces) still reach memlogd and then end up in cloud/ES.

eriknordmark commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -215,6 +258,20 @@ func printToFile(filename string, str string) error {
 
+// Overwrite if file exists.
+func overWriteFile(filename string, str string) error {
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE, 0755)
```
Add O_TRUNC

gkodali-zededa replied:
added

rvs commented:
> Doesn't mean we shouldn't make the logging pipeline reliable and usable end to end, but that seems to still be months away.

Fair enough @eriknordmark -- I am still unhappy about `spoofStdFDs` though -- the way I read it seems that we're now circumventing the normal stdout/stderr flow that feeds memlogd? How's that supposed to work @gkodali-zededa ?
